### ISSUE TITLE
Sysl Parser and Proto Structure Proposals

### ISSUE THREAD

**Issue by cuminandpaprika** [anz-bank/sysl#874]
This is an issue to capture any issues and ideas for improvements for the way sysl is parsed and stored in a data structure.

## Context

Currently, the data model has evolved through various feature requests and become very complex, with some inconsistencies around where and how data is stored. This makes developing code which interacts with the Sysl data structure very complex and verbose, needing to handle a variety of cases.

To speed up development of sysl features, a number of improvements can be made to the data structure, as well as the parser (when the parser is migrated in the future from ANTLR to WBNF) https://github.com/anz-bank/sysl/issues/581.

### Endpoint Statements
Return statements are stored as strings rather than structured data, requiring any call statements and return statements to be parsed by feature code, rather than being handled in the parser itself.

### Type System

A number of types defined in `sysl.pb.go` are not used or duplications. 
- Sequence and List

### Names
```
My App:
    ...
```
Currently names allow for spaces in the definition, however this is rarely used and most users define names as an _UpperCamelCase_ string instead, and a LongName in square braces
```
MyApp["My Long Application Name"]:
   ...
```

### Namespaces

```
Namespace1::Namespace2::MyApp:
   ...
```
Sysl also provides a namespacing syntax which is not widely used. This namespacing means that all application names are stored as an array of strings, requiring sysl developers to concatenate the string in almost all use cases.

### Type Reference consistency

The type reference system is highly complex, with very little documentation regarding the behaviour. Most applications need to parse these and provide additional information from the context for references within the same app.

For example, the following:
```
App1:
    someEndpoint(myParam <: myType):
        return string

    !type myType:
        content <: string
```

is parsed into the following data structure:
```
param:  {
          name:  "myParam"
          type:  {
            type_ref:  {
              ref:  {
                appname:  {
                  part:  "myType"
                }
              }
            }
          }
        }
```
The TypeName is found in the AppName field, which is incorrect. More details here https://github.com/anz-bank/sysl/issues/817. This bug is currently handled in sysl-catalog, but should be fixed in the source.

In addition, the param fails to consistently populate the source and context fields, making it very difficult to parse parameters consistently.

### Parameter Inconsistency

Currently, Path, Query, Response and Header parameters are stored in an inconsistent manner, which each being a different type. The lack of consistency means that new developers find it hard to understand, from the data-structure alone where to find a parameter when working with the Sysl data type.

### Custom Patterns

A number of patterns exist to work around limitations in the type system. E.g ~empty, ~jsontag etc. are all used to indicate information which is unhandled by the parser. These should be documented in one place, so that a user can see exactly what each pattern means in each context.

These are just some pain points found when working, to inform the refactoring to be done when we migrate to the WBNF parser

**Comment by nydrani** [comment]
Endpoint Statements: These can probably be enhanced but some statements are just raw string statements. I think we can have specific syntax for return statements (or statements that return some kind of data) for easier parsing.

Names: I didnt know that existed.

I do agree that the sysl data structure is tedious to access and parse, If we were to consolidate the data structure, it would probably mean backwards compatability is required. For alot of the pain points, it kind of looks like many of issues could be a symptom of being afterthoughts in the design of sysl.

**Comment by cuminandpaprika** [comment]
> Endpoint Statements: These can probably be enhanced but some statements are just raw string statements. I think we can have specific syntax for return statements (or statements that return some kind of data) for easier parsing.

Yeah I think we can still allow for any raw strings, but have a formal syntax for call and return statements. e.g we haven't really documented that the AppName can be dropped when referencing types defined in the same application

> I do agree that the sysl data structure is tedious to access and parse, If we were to consolidate the data structure, it would probably mean backwards compatability is required. For alot of the pain points, it kind of looks like many of issues could be a symptom of being afterthoughts in the design of sysl.

Yeah lots of the issues seem to be due to the nature of Sysl having evolved over time to suit various needs. Could probably do it two ways:

1. Maintain backwards compatability and make the changes that don't break the grammar to the `sysl.pb.go` struct. Other changes might need to be made using something like `syslwrapper`
2. Release Sysl2.0 with some automation for migrating to the new language version. Would probably want to put a lot of thought into what this version of Sysl would look like since we only want to do it once. 
## Sysl Proto Struct
An additional question I have is what benefits are offered by using a struct generated from a proto file for sysl. Do these benefits outweigh the verboseness and rigidity enforced by the proto-generated code? (Just a bit confused about the original design intent)
